### PR TITLE
Complete RTS compiler args

### PR DESCRIPTION
**Pull request by izgzhen**
solve https://github.com/snowleopard/hadrian/issues/406

### CODE REVIEW

**Comment by izgzhen** [comment]
@snowleopard you mentioned its relevance with validation errors -- so I will try to get some validate result at the same time when you review this code as well.

**Comment by izgzhen** [comment]
current master:

```
SUMMARY for test run started at Thu Sep 21 20:15:33 2017 HKT
 0:08:11 spent to go through
    6078 total tests, which gave rise to
   14959 test cases, of which
    9003 were skipped

      40 had missing libraries
    5449 expected passes
     158 expected failures

       0 caused framework failures
       0 caused framework warnings
       3 unexpected passes
     223 unexpected failures
      83 unexpected stat failures
```

this patch:

```
SUMMARY for test run started at Thu Sep 21 19:37:30 2017 HKT
 0:06:10 spent to go through
    6078 total tests, which gave rise to
   14959 test cases, of which
    9003 were skipped

      40 had missing libraries
    5448 expected passes
     158 expected failures

       0 caused framework failures
       0 caused framework warnings
       3 unexpected passes
     224 unexpected failures
      83 unexpected stat failures
```

Not much difference though

**Comment by izgzhen** [comment]
BTW, I was working around the ambiguous `RuleResult` error in CI at another branch. Weird that I can't reproduce this on my machine -- is it caused by different compiler version?

**Comment by izgzhen** [comment]
Oh .. I guess it is because Shake has just updated new version 0.16 and the CI build includes it.

File: src/Settings/Packages/Rts.hs [hunk 1]
```diff
@@ -29,13 +38,67 @@ rtsLibffiLibraryName = do
...
     suf     <- libsuf way
     rtsPath <- rtsBuildPath
     return $ rtsPath -/- "lib" ++ name ++ suf
 
+-- ref: mk/config.mk.in
+ghcRtsWithLibDw :: Action Bool
+ghcRtsWithLibDw = do
+    goodArch <- anyTargetArch ["i386", "x86_64"]
+    withLibDw <- setting WithLibDw
+    return $ goodArch && withLibDw == "YES"
```
Comment (snowleopard) [inline src/Settings/Packages/Rts.hs]: It looks like `WithLibDw` should be a `Flag`, not a `Setting`, so the comparison with `"YES"` is taken care of automatically.

File: src/Settings/Packages/Rts.hs [hunk 2]
```diff
@@ -122,137 +212,7 @@ rtsPackageArgs = package rts ? do
...
--- # for details
-
--- # Without this, thread_obj will not be inlined (at least on x86 with GCC 4.1.0)
--- ifneq "$(CC_CLANG_BACKEND)" "1"
--- rts/sm/Compact_CC_OPTS += -finline-limit=2500
--- endif
+               , "-DINCLUDE_DIR=\"" ++ destDir ++ ghclibDir -/- "include\"" ]
+        , builder HsCpp ? mconcat
+            [ ghcRtsWithLibDw ? arg "-DUSE_LIBDW"
+            , (== "YES") <$> setting HaveLibMingwEx ? arg "-DHAVE_LIBMINGWEX" ] ]
```
Comment (snowleopard) [inline src/Settings/Packages/Rts.hs]: `HaveLibMingwEx` should also be a `Flag`, not a `Setting`.

File: src/Settings/Packages/Rts.hs [hunk 3]
```diff
@@ -29,13 +38,67 @@ rtsLibffiLibraryName = do
...
 
+-- ref: mk/config.mk.in
+ghcRtsWithLibDw :: Action Bool
+ghcRtsWithLibDw = do
+    goodArch <- anyTargetArch ["i386", "x86_64"]
+    withLibDw <- setting WithLibDw
+    return $ goodArch && withLibDw == "YES"
+
+ccClangBackend :: Action Bool
+ccClangBackend = (== "1") <$> setting CcClangBackend
```
Comment (snowleopard) [inline src/Settings/Packages/Rts.hs]: Ohh, is `"1"` here equivalent to `"YES"`? Or is this a version number?

**Comment by snowleopard** [comment]
@izgzhen Great, thanks for clearing this up! I've left a couple of suggestions to keep the comparison of flag-like settings to `"YES"` in one place.

File: src/Settings/Packages/Rts.hs (hunk 3 above)
Comment (snowleopard) [reply src/Settings/Packages/Rts.hs]: If this is a version, I suggest we rename this to something like `ccClangBackendVersionEq1` to distinguish it from YES/NO settings.

File: src/Settings/Packages/Rts.hs (hunk 1 above)
Comment (izgzhen) [reply src/Settings/Packages/Rts.hs]: Oh I see, thanks!

File: src/Settings/Packages/Rts.hs (hunk 3 above)
Comment (izgzhen) [reply src/Settings/Packages/Rts.hs]: I think it is using "1" to represent "YES", and "0" for "NO".  Can we simply use `GccIsClang` here?

Comment (izgzhen) [reply src/Settings/Packages/Rts.hs]: See https://github.com/ghc/ghc/blob/master/aclocal.m4#L1250

Comment (snowleopard) [reply src/Settings/Packages/Rts.hs]: @izgzhen Yes, it looks like we can use `GccIsClang` instead. Perhaps we could even drop `CcClangBackend` altogether?

Comment (izgzhen) [reply src/Settings/Packages/Rts.hs]: Indeed, I was thinking about doing that

**Comment by izgzhen** [comment]
cleaned up the bits mentioned

**Comment by snowleopard** [comment]
Thanks @izgzhen! Merged.

**Pull request merged**